File: projects/ludo/stack-springai/src/main/java/com/llmeval/ludo/springai/Prompts.java
package com.llmeval.ludo.springai;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.yaml.snakeyaml.Yaml;

/**
 * Loads the shared prompt set — the Java half of the contract prompts.py keeps
 * in Python. Both must agree exactly: same validation, same literal
 * {@code {{name}}} rendering, same digest bytes, or the "identical prompts"
 * claim quietly stops being true.
 *
 * <p>No template logic, deliberately: two languages implementing conditionals
 * would disagree at the edges, and every disagreement would be an invisible
 * parity break (shared/prompts/README.md). Rendering is ten lines in either
 * language — that is the point.
 */
public final class Prompts {

    private static final Pattern VARIABLE = Pattern.compile("\\{\\{(\\w+)}}");
    private static final Pattern LOGIC =
            Pattern.compile("\\{\\{[#/^>!]|\\{%|\\$\\{|\\{\\{\\s*\\w+\\s*\\|");

    public record Template(String name, String text, List<String> variables) {

        /** Substitute every declared variable. Missing or extra ones are errors. */
        public String render(Map<String, Object> values) {
            Set<String> missing = new HashSet<>(variables);
            missing.removeAll(values.keySet());
            if (!missing.isEmpty()) {
                throw new IllegalArgumentException(name + ": missing " + missing);
            }
            Set<String> extra = new HashSet<>(values.keySet());
            variables.forEach(extra::remove);
            if (!extra.isEmpty()) {
                throw new IllegalArgumentException(name + ": undeclared " + extra);
            }
            String out = text;
            for (Map.Entry<String, Object> entry : values.entrySet()) {
                out = out.replace("{{" + entry.getKey() + "}}", String.valueOf(entry.getValue()));
            }
            return out;
        }
    }

    private final int version;
    private final String digest;
    private final List<Template> system;
    private final Map<String, Template> turn;

    private Prompts(int version, String digest, List<Template> system, Map<String, Template> turn) {
        this.version = version;
        this.digest = digest;
        this.system = system;
        this.turn = turn;
    }

    public int version() { return version; }

    /** {@code sha256:…} over the manifest and every template — same bytes as prompts.py. */
    public String digest() { return digest; }

    public Template turn(String name) {
        Template t = turn.get(name);
        if (t == null) throw new IllegalArgumentException("no turn template " + name);
        return t;
    }

    public Set<String> turnNames() { return turn.keySet(); }

    /** The cacheable layer: every {@code system/} template, concatenated in order. */
    public String systemPrompt(Map<String, Object> values) {
        List<String> parts = new ArrayList<>();
        for (Template t : system) {
            Map<String, Object> narrowed = new HashMap<>();
            for (String v : t.variables()) narrowed.put(v, values.get(v));
            parts.add(t.render(narrowed));
        }
        return String.join("\n\n", parts);
    }

    /** Walk up from the working directory to the repo root, where shared/ lives. */
    public static Path repoRoot() {
        Path here = Path.of("").toAbsolutePath();
        for (Path p = here; p != null; p = p.getParent()) {
            if (Files.exists(p.resolve("shared/prompts/ludo/manifest.yaml"))) return p;
        }
        throw new IllegalStateException("could not locate shared/prompts/ludo from " + here);
    }

    public static Prompts load() {
        return load(repoRoot().resolve("shared/prompts/ludo"));
    }

    @SuppressWarnings("unchecked")
    public static Prompts load(Path base) {
        try {
            Map<String, Object> manifest =
                    new Yaml().load(Files.readString(base.resolve("manifest.yaml"), StandardCharsets.UTF_8));
            int version = (int) manifest.get("version");

            List<Map<String, Object>> systemEntries = (List<Map<String, Object>>) manifest.get("system");
            Map<String, Map<String, Object>> turnEntries =
                    (Map<String, Map<String, Object>>) manifest.get("turn");

            List<Template> system = new ArrayList<>();
            for (Map<String, Object> entry : systemEntries) system.add(loadTemplate(base, entry));
            Map<String, Template> turn = new LinkedHashMap<>();
            turnEntries.forEach((name, entry) -> turn.put(name, loadTemplate(base, entry)));

            return new Prompts(version, digest(base, version, systemEntries, turnEntries), system, turn);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    @SuppressWarnings("unchecked")
    private static Template loadTemplate(Path base, Map<String, Object> entry) {
        String file = (String) entry.get("file");
        String text;
        try {
            text = Files.readString(base.resolve(file), StandardCharsets.UTF_8);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        if (LOGIC.matcher(text).find()) {
            throw new IllegalStateException(file + ": template logic found — render it in code "
                    + "and pass the result in as one variable (shared/prompts/README.md)");
        }
        List<String> declared = entry.get("variables") == null
                ? List.of() : (List<String>) entry.get("variables");
        Set<String> used = new HashSet<>();
        Matcher m = VARIABLE.matcher(text);
        while (m.find()) used.add(m.group(1));
        if (!used.equals(new HashSet<>(declared))) {
            throw new IllegalStateException(file + ": manifest declares " + declared
                    + ", template uses " + used);
        }
        return new Template(file, text, declared);
    }

    /** Byte-for-byte the prompts.py algorithm: version, then each entry's file
     *  name and bytes, NUL-separated, manifest order. */
    private static String digest(Path base, int version, List<Map<String, Object>> system,
                                 Map<String, Map<String, Object>> turn) throws IOException {
        MessageDigest sha;
        try {
            sha = MessageDigest.getInstance("SHA-256");
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException(e);
        }
        sha.update(String.valueOf(version).getBytes(StandardCharsets.UTF_8));
        List<Map<String, Object>> entries = new ArrayList<>(system);
        entries.addAll(turn.values());
        for (Map<String, Object> entry : entries) {
            String file = (String) entry.get("file");
            sha.update(file.getBytes(StandardCharsets.UTF_8));
            sha.update((byte) 0);
            sha.update(Files.readAllBytes(base.resolve(file)));
            sha.update((byte) 0);
        }
        StringBuilder hex = new StringBuilder("sha256:");
        for (byte b : sha.digest()) hex.append(String.format("%02x", b));
        return hex.toString();
    }
}

File: projects/ludo/stack-langgraph/src/ludo_langgraph/prompts.py
"""Loading the shared prompt set.

Every stack loads the same files from ``shared/prompts/ludo`` and sends them
verbatim. This module is this stack's copy of that contract — stacks cannot
share code (a shortcut between them would destroy the comparison), so each
carries its own loader and the digest proves they agree byte for byte.

The whole template language is ``{{name}}`` -> string. No conditionals, no
loops, no filters — see shared/prompts/README.md for why: template logic would
have to be implemented identically in two languages, and the places they
disagreed would be silent parity breaks in the one file whose job is to
guarantee parity.
"""

from __future__ import annotations

import hashlib
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import yaml

#: Anything a template engine would treat as control flow. Rendering must fail
#: loudly rather than quietly emit it, or a stack could smuggle logic in.
LOGIC = re.compile(r"\{\{[#/^>!]|\{%|\$\{|\{\{\s*\w+\s*\|")

VARIABLE = re.compile(r"\{\{(\w+)\}\}")


def repo_root() -> Path:
    """Walk up to the repository root, which is where ``shared/`` lives."""
    here = Path(__file__).resolve()
    for parent in here.parents:
        if (parent / "shared" / "prompts" / "ludo" / "manifest.yaml").exists():
            return parent
    raise FileNotFoundError("could not locate shared/prompts/ludo from " + str(here))


@dataclass(frozen=True)
class Template:
    name: str
    text: str
    variables: tuple[str, ...]

    def render(self, **values: Any) -> str:
        """Substitute every declared variable. Missing or extra ones are errors.

        Strict on purpose. A silently unsubstituted ``{{board}}`` would reach a
        model as literal braces and produce plausible-looking nonsense; a
        stack quietly passing an undeclared variable would be a divergence the
        other two stacks do not have.
        """
        missing = set(self.variables) - set(values)
        if missing:
            raise KeyError(f"{self.name}: missing {sorted(missing)}")
        extra = set(values) - set(self.variables)
        if extra:
            raise KeyError(f"{self.name}: undeclared {sorted(extra)}")

        text = self.text
        for key, value in values.items():
            text = text.replace("{{" + key + "}}", str(value))
        return text


@dataclass(frozen=True)
class PromptSet:
    """The loaded prompt set, plus the provenance a transcript records."""

    version: int
    digest: str
    system: tuple[Template, ...]
    turn: dict[str, Template]

    def system_prompt(self, **values: Any) -> str:
        """The cacheable layer: every ``system/`` template, concatenated in order.

        Stable for the whole game, which is what makes it prompt-cacheable.
        Anything that changes per turn belongs in a ``turn/`` template instead.
        """
        return "\n\n".join(t.render(**{k: values[k] for k in t.variables})
                           for t in self.system)

    def provenance(self) -> dict[str, Any]:
        """The ``game_started.prompt_set`` payload."""
        return {"version": self.version, "hash": self.digest}


def _load_template(base: Path, entry: dict[str, Any]) -> Template:
    path = base / entry["file"]
    text = path.read_text(encoding="utf-8")

    if LOGIC.search(text):
        raise ValueError(
            f"{entry['file']}: template logic found. Render it in code and pass "
            f"the result in as one variable — see shared/prompts/README.md"
        )

    declared = tuple(entry.get("variables") or ())
    used = set(VARIABLE.findall(text))
    if used != set(declared):
        raise ValueError(
            f"{entry['file']}: manifest declares {sorted(declared)}, "
            f"template uses {sorted(used)}"
        )
    return Template(entry["file"], text, declared)


def load(base: Path | None = None) -> PromptSet:
    """Load and validate the prompt set, and compute its digest."""
    base = base or (repo_root() / "shared" / "prompts" / "ludo")
    manifest = yaml.safe_load((base / "manifest.yaml").read_text(encoding="utf-8"))

    system = tuple(_load_template(base, e) for e in manifest["system"])
    turn = {name: _load_template(base, e) for name, e in manifest["turn"].items()}

    return PromptSet(
        version=manifest["version"],
        digest=digest(base, manifest),
        system=system,
        turn=turn,
    )


def digest(base: Path, manifest: dict[str, Any]) -> str:
    """Content hash over the manifest and every template it names.

    Recorded in ``game_started`` so a transcript names exactly the prompts that
    produced it. A version number alone can be forgotten on the way out the
    door; a hash cannot. Order is taken from the manifest rather than from the
    filesystem, so two checkouts hash the same regardless of directory order.
    """
    h = hashlib.sha256()
    h.update(str(manifest["version"]).encode("utf-8"))
    entries = list(manifest["system"]) + list(manifest["turn"].values())
    for entry in entries:
        h.update(entry["file"].encode("utf-8"))
        h.update(b"\0")
        h.update((base / entry["file"]).read_bytes())
        h.update(b"\0")
    return "sha256:" + h.hexdigest()
